src/libsystemd-network/sd-ndisc.c

```diff
@@ -420,6 +421,277 @@ _public_ int sd_ndisc_start(sd_ndisc *nd) {
+                zero(ra->mac_addr);
+
+        return 0;
+}
+
+_public_ int sd_radv_set_mtu(sd_radv *ra, uint32_t mtu) {
+        assert_return(ra, -EINVAL);
+        assert_return(mtu >= 1280, -EINVAL);
+
+        if (ra->state != SD_RADV_STATE_IDLE)
+                return -EBUSY;
+
+        ra->mtu = mtu;
+
+        return 0;
+}
+
+_public_ int sd_radv_set_hop_limit(sd_radv *ra, uint8_t hop_limit) {
+        assert_return(ra, -EINVAL);
+
+        if (ra->state != SD_RADV_STATE_IDLE)
+                return -EBUSY;
+
+        ra->hop_limit = hop_limit;
+
+        return 0;
+}
+
+_public_ int sd_radv_set_router_lifetime(sd_radv *ra, uint32_t router_lifetime) {
+        assert_return(ra, -EINVAL);
+
+        if (ra->state != SD_RADV_STATE_IDLE)
+                return -EBUSY;
+
+        ra->lifetime = router_lifetime;
+
+        if (ra->lifetime == 0)
+                (void) sd_radv_set_preference(ra, SD_NDISC_PREFERENCE_MEDIUM);
+
+        return 0;
+}
+
+_public_ int sd_radv_set_managed_information(sd_radv *ra, bool managed) {
+        assert_return(ra, -EINVAL);
+
+        if (ra->state != SD_RADV_STATE_IDLE)
+                return -EBUSY;
+
+        if (managed)
+                ra->flags |= ND_RA_FLAG_MANAGED;
+        else
+                ra->flags &= ~ND_RA_FLAG_MANAGED;
+
+        return 0;
+}
+
+_public_ int sd_radv_set_other_information(sd_radv *ra, bool other) {
+        assert_return(ra, -EINVAL);
+
+        if (ra->state != SD_RADV_STATE_IDLE)
+                return -EBUSY;
+
+        if (other)
+                ra->flags |= ND_RA_FLAG_OTHER;
+        else
+                ra->flags &= ~ND_RA_FLAG_OTHER;
+
+        return 0;
+}
+
+_public_ int sd_radv_set_preference(sd_radv *ra, unsigned preference) {
+        int r = 0;
+
+        assert_return(ra, -EINVAL);
+        assert_return(IN_SET(preference,
+                             SD_NDISC_PREFERENCE_LOW,
+                             SD_NDISC_PREFERENCE_MEDIUM,
+                             SD_NDISC_PREFERENCE_HIGH), -EINVAL);
+
+        /* RFC 4191, Section 2.2, "...If the Router Lifetime is zero, the
+           preference value MUST be set to (00) by the sender..." */
+        if (ra->lifetime == 0) {
+                preference = SD_NDISC_PREFERENCE_MEDIUM;
+                r = -EINVAL;
+        }
+
+        ra->flags = (ra->flags & ~(0x3 << 3)) | (preference << 3);
+
+        return r;
+}
+
+_public_ int sd_radv_add_prefix(sd_radv *ra, sd_radv_prefix *p) {
+        sd_radv_prefix *cur;
+        char buffer[INET6_ADDRSTRLEN];
+
+        assert_return(ra, -EINVAL);
+
+        if (!p)
+                return -EINVAL;
+
+        LIST_FOREACH(prefix, cur, ra->prefixes) {
+                unsigned prefixlen, bits, bytes, mask, byte_cur, byte_p;
+
+                prefixlen = MIN(cur->opt.prefixlen, p->opt.prefixlen);
+                bytes = prefixlen / 8;
+                bits = prefixlen % 8;
+
+                if (memcmp(&cur->opt.in6_addr, &p->opt.in6_addr, bytes))
```

Comment:
` memcmp() != 0` please (we're not entirely consistent, but the general rule is to use implicit casts to bool only for pointers and bools).